docs/design/velero-krm-enhancement.md

```diff
@@ -0,0 +1,103 @@
+---
+title: velero-openshiftkrm-plugin design
+authors:
+  - "@ChaitanyaBhangale"
+reviewers:
+  - "@jwmatthews" 
+  - "@shubham-pampattiwar"
+  - "@dylan-murray"
+  - "@tiger-kaovilai"
+approvers:
+  - "@jwmatthews" 
+  - "@dylan-murray"
+  - "@shubham-pampattiwar"
+creation-date: 2022-06-21
+last-updated: 2022-06-21
+status: 
+  - "N/A"
+see-also:
+  - "N/A"  
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# Velero KRM Plugin Design
+
+### Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] User-facing documentation is created
+
+### Open Questions 
+-  Can changes be made in exising codebase as opposed to refactoring the whole code? 
+-  Whether adding KRM (Kubernetes Resource Management) functions is actually useful or will it affect the performance of current plugin in a negative manner?
+
+
+### Summary
+
+Currently Velero plugin framework is inefficient. Velero server executes 1 gRPC call per plugin, meaning 1 gRPC call per resource transformation.  In order to overcome this we introduce a new plugin type that will allow Velero to leverage KRM functions for resource transformation and additional item gathering so that we can drastically improve the plugin framework performance by processing N items with a single gRPC call.
+
+### Motivation
+
+- To have a common set of KRM functions which are reusable.
+- Improving the performance of Velero backup-restore process by only having one grpc call per plugin when processing backup/restore of a resource.
+and passing a list of resources as opposed to having one grpc call to backup/restore a resource. 
+
+### Goals
+- Check and research on if the given approach is feasible.
+- Make changes to OpenShift Velero Plugins and introduce a new Velero plugin type to which accepts a resourceList and uses KRM functions internally.
+- Modify our velero fork to pass a resource list to a new velero plugin interface
+
+### User Stories 
+- User will be developing a new plugin type that can handle multiple resources for common transformations
+- User will be able to pass all resource list in velero plugin interface as compared to the current one to one mapping of resources 
+
+Currently openshift velero plugin interface can accept a single input in the input parameter for the given code snippet. A new plugin type will be added which will accept the input as a list of resources and the modified snippet will look like
+```
+func (p *RestorePlugin) Execute(input *velero.RestoreItemActionExecuteInput) 
+(*velero.RestoreItemActionExecuteOutput, error) {
+
+	// KRM function logic to transform a particular resource
+        // Convert input to KubeObject for resource transformation 
+
+}
+```
+Here input will be a list of resources, output will be tranformed resource and the function will consist of KRM logic
+
+ 
+## Design and Implementation Details
```

Comment:
Something about the phase's and the two parts of phase one are just not lining up for me. 

Phase 1 is an investigation, of if we can change the current OpenshiftVeleroPlugin to run a KRM function, that only has a single resource in the resourcelist. 

Phase 2 is two things, building a new plugin based on the new plugin type that we will add to our fork.

That is what I think I am reading can you help me understand?

Reply:
So phase 1 is just identifying if we can add KRM functions directly in the current codebase without adding a new plugin and Phase 2 is adding a new plugin which accepts a list of resources in the velero-plugins repo code and modifying the controller to pass in a list of resources to the given plugin